runtime/lua/vim/lsp/folding_range.lua

```diff
@@ -0,0 +1,417 @@
+        ---@param client vim.lsp.Client
+        :filter(function(client)
+          return client.id ~= client_id
+        end)
+        :totable()
+      if #clients == 0 then
+        if api.nvim_buf_is_loaded(bufnr) then
+          bufstates[bufnr] = {
+            client_ranges = {},
+            row_level = {},
+            row_kinds = {},
+            row_text = {},
+          }
+          foldupdate(bufnr)
+        end
+      else
+        for _, client in ipairs(clients) do
+          request(bufnr, client)
+        end
+      end
+    end,
+  })
+  api.nvim_create_autocmd('LspAttach', {
+    group = augroup_setup,
+    buffer = bufnr,
+    callback = function(args)
+      local client = assert(vim.lsp.get_client_by_id(args.data.client_id))
+      request(bufnr, client)
+    end,
+  })
+  api.nvim_create_autocmd('LspNotify', {
+    group = augroup_setup,
+    buffer = bufnr,
+    callback = function(args)
+      local client = assert(vim.lsp.get_client_by_id(args.data.client_id))
+      if
+        client:supports_method(ms.textDocument_foldingRange, bufnr)
+        and (
+          args.data.method == ms.textDocument_didChange
+          or args.data.method == ms.textDocument_didOpen
+        )
+      then
+        request(bufnr, client)
+      end
+    end,
+  })
+
+  for _, client in
+    ipairs(vim.lsp.get_clients({ bufnr = bufnr, method = ms.textDocument_foldingRange }))
+  do
+    request(bufnr, client)
+  end
+
+  return bufstates[bufnr]
+end
+
+-- The "garbage collector" for `bufstate` and its associated event hooks.
+-- See the comment above `augroup_setup` function for more details.
+--
+-- It will be fine without this, as the bufstate and event hooks will also
+-- be cleared when the buffer is unloaded.
+-- api.nvim_create_autocmd('OptionSet', {
+--   group = augroup_setup,
+--   pattern = 'foldexpr',
+--   callback = function()
+--     -- Sould find a more reliable way to detect changes in `foldexpr`.
+--     if vim.v.option_old == 'v:lua.vim.lsp.foldexpr()' then
+--       for bufnr in pairs(bufstates) do
+--         if
+--           ---@param winid integer
+--           vim.iter(vim.fn.win_findbuf(bufnr)):all(function(winid)
+--             return vim.wo[winid].foldexpr ~= 'v:lua.vim.lsp.foldexpr()'
+--           end)
+--         then
+--           bufstates[bufnr] = nil
+--           api.nvim_clear_autocmds({ buffer = bufnr, group = augroup_setup })
+--         end
+--       end
+--     end
+--   end,
+-- })
+
+---@param kind lsp.FoldingRangeKind
+---@param winid integer
+local function foldclose(kind, winid)
+  vim._with({ win = winid }, function()
+    local bufnr = api.nvim_win_get_buf(winid)
+    local row_kinds = bufstates[bufnr].row_kinds
+    -- Reverse traverse to ensure that the smallest ranges are closed first.
+    for row = api.nvim_buf_line_count(bufnr) - 1, 0, -1 do
+      local kinds = row_kinds[row]
+      if kinds and kinds[kind] then
+        vim.cmd(row + 1 .. 'foldclose')
+      end
```

Comment:
This looks quite costly. Doesn't `vim.cmd.normal('zM')` do this?

Reply:
Thanks, I didn't know there were performance differences in this. But I think it is still necessary to use `:foldclose`, because `zc` affects `foldenable` and `'zm'` affects `'foldlevel'`, we should choose one that has no side effects.

Reply:
Running a command on every single line in the buffer will exhibit issues on large files.

I thought I saw `foldclose` called in other places in the `foldexpr` path, but I can't see any now, so `foldclose` looks like it can only be triggered directly by the user, so will be ok for now.

Reply:
I have precautions regarding this. `:foldclose` will not be called on every line, but rather at the beginning of a certain `kind`. If it is folding imports, it will only be called once; if it is folding comments, it will also only be called on the first line so even if a file has 100 functions, each with a docstring, it will only be called 100 times. This should already be quite large.